activesupport/lib/active_support/messages/metadata.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+require "time"
+
+module ActiveSupport
+  module Messages
+    class Metadata #:nodoc:
+      def initialize(expires_at, purpose)
+        @expires_at, @purpose = expires_at, purpose
+      end
+
+      class << self
+        def wrap(message, expires_at: nil, expires_in: nil, purpose: nil)
+          { "value" => message, "_rails" => { "exp" => pick_expiry(expires_at, expires_in), "pur" => purpose } }
+        end
+
+        def verify(message, purpose)
+          metadata = extract_metadata(message)
+          if metadata.nil?
+            message
+          elsif metadata.match?(purpose) && metadata.fresh?
+            message["value"]
+          else
+            nil
+          end
+        end
+
+        private
+          def pick_expiry(expires_at, expires_in)
+            if expires_at
+              expires_at.utc.iso8601(3)
+            elsif expires_in
+              expires_in.from_now.utc.iso8601(3)
+            else
+              nil
+            end
+          end
+
+          def extract_metadata(message)
+            if message.is_a?(Hash) && message.key?("_rails")
+              metadata = new(message["_rails"]["exp"], message["_rails"]["pur"])
+              return metadata
```

Comment:
Ruby automatically returns the last evaluated statement, so unneeded return statements are generally frowned upon — and we don't need the `metadata` variable here either.